// @terrestris/react-geo/dist/Grid/AgFeatureGrid/AgFeatureGrid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.AgFeatureGrid = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _agGridReact = require("ag-grid-react");

var _Style = _interopRequireDefault(require("ol/style/Style"));

var _Fill = _interopRequireDefault(require("ol/style/Fill"));

var _Circle = _interopRequireDefault(require("ol/style/Circle"));

var _Stroke = _interopRequireDefault(require("ol/style/Stroke"));

var _Map = _interopRequireDefault(require("ol/Map"));

var _Feature = _interopRequireDefault(require("ol/Feature"));

var _Vector = _interopRequireDefault(require("ol/source/Vector"));

var _Vector2 = _interopRequireDefault(require("ol/layer/Vector"));

var _Geometry = _interopRequireDefault(require("ol/geom/Geometry"));

var _GeometryCollection = _interopRequireDefault(require("ol/geom/GeometryCollection"));

var _isArray = _interopRequireDefault(require("lodash/isArray"));

var _differenceWith = _interopRequireDefault(require("lodash/differenceWith"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _isFunction = _interopRequireDefault(require("lodash/isFunction"));

var _kebabCase = _interopRequireDefault(require("lodash/kebabCase"));

var _MapUtil = _interopRequireDefault(require("@terrestris/ol-util/dist/MapUtil/MapUtil"));

var _constants = require("../../constants");

require("ag-grid-community/dist/styles/ag-grid.css");

require("ag-grid-community/dist/styles/ag-theme-balham.css");

require("ag-grid-community/dist/styles/ag-theme-fresh.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { if (i % 2) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } else { Object.defineProperties(target, Object.getOwnPropertyDescriptors(arguments[i])); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The AgFeatureGrid.
 *
 * @class The AgFeatureGrid
 * @extends React.Component
 */
var AgFeatureGrid =
/*#__PURE__*/
function (_React$Component) {
  _inherits(AgFeatureGrid, _React$Component);

  /**
   * The className added to this component.
   * @type {String}
   * @private
   */

  /**
   * The class name to add to each table row.
   * @type {String}
   * @private
   */

  /**
   * The prefix to use for each table row class.
   * @type {String}
   * @private
   */

  /**
   * The hover class name.
   * @type {String}
   * @private
   */

  /**
   * The source holding the features of the grid.
   * @type {ol.source.Vector}
   * @private
   */

  /**
   * The layer representing the features of the grid.
   * @type {ol.layer.Vector}
   * @private
   */

  /**
   * The properties.
   * @type {Object}
   */

  /**
   * The default properties.
   * @type {Object}
   */
  function AgFeatureGrid(_props) {
    var _this;

    _classCallCheck(this, AgFeatureGrid);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AgFeatureGrid).call(this, _props));

    _defineProperty(_assertThisInitialized(_this), "_className", "".concat(_constants.CSS_PREFIX, "ag-feature-grid"));

    _defineProperty(_assertThisInitialized(_this), "_rowClassName", "".concat(_constants.CSS_PREFIX, "ag-feature-grid-row"));

    _defineProperty(_assertThisInitialized(_this), "_rowKeyClassNamePrefix", 'row-key-');

    _defineProperty(_assertThisInitialized(_this), "_rowHoverClassName", 'ag-row-hover');

    _defineProperty(_assertThisInitialized(_this), "_source", null);

    _defineProperty(_assertThisInitialized(_this), "_layer", null);

    _defineProperty(_assertThisInitialized(_this), "initVectorLayer", function (map) {
      var _this$props = _this.props,
          features = _this$props.features,
          featureStyle = _this$props.featureStyle,
          layerName = _this$props.layerName;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      if (_MapUtil["default"].getLayerByName(map, layerName)) {
        return;
      }

      var source = new _Vector["default"]({
        features: features
      });
      var layer = new _Vector2["default"]({
        name: layerName,
        source: source,
        style: featureStyle
      });
      map.addLayer(layer);
      _this._source = source;
      _this._layer = layer;
    });

    _defineProperty(_assertThisInitialized(_this), "initMapEventHandlers", function (map) {
      var selectable = _this.props.selectable;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      map.on('pointermove', _this.onMapPointerMove);

      if (selectable) {
        map.on('singleclick', _this.onMapSingleClick);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onMapPointerMove", function (olEvt) {
      var _this$props2 = _this.props,
          map = _this$props2.map,
          features = _this$props2.features,
          highlightStyle = _this$props2.highlightStyle,
          selectStyle = _this$props2.selectStyle;
      var grid = _this.state.grid;

      var selectedRowKeys = _this.getSelectedRowKeys();

      var selectedFeatures = map.getFeaturesAtPixel(olEvt.pixel, {
        layerFilter: function layerFilter(layerCand) {
          return layerCand === _this._layer;
        }
      }) || [];

      if (!grid || !grid.api) {
        return;
      }

      var rowRenderer = grid.api.rowRenderer;
      features.forEach(function (feature) {
        var key = _this.props.keyFunction(feature);

        var rc;
        rowRenderer.forEachRowComp(function (idx, rowComp) {
          if (rowComp.getRowNode().data.key === key) {
            rc = rowComp;
          }
        });

        if (rc) {
          var el = rc.getBodyRowElement();

          if (el) {
            el.classList.remove(_this._rowHoverClassName);
          }
        }

        if (selectedRowKeys.includes(key)) {
          feature.setStyle(selectStyle);
        } else {
          feature.setStyle(null);
        }
      });
      selectedFeatures.forEach(function (feature) {
        var key = _this.props.keyFunction(feature);

        var rc;
        rowRenderer.forEachRowComp(function (idx, rowComp) {
          if (rowComp.getRowNode().data.key === key) {
            rc = rowComp;
          }
        });

        if (rc) {
          var el = rc.getBodyRowElement();

          if (el) {
            el.classList.add(_this._rowHoverClassName);
          }
        }

        feature.setStyle(highlightStyle);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onMapSingleClick", function (olEvt) {
      var _this$props3 = _this.props,
          map = _this$props3.map,
          selectStyle = _this$props3.selectStyle,
          onMapSingleClick = _this$props3.onMapSingleClick;

      var selectedRowKeys = _this.getSelectedRowKeys();

      var selectedFeatures = map.getFeaturesAtPixel(olEvt.pixel, {
        layerFilter: function layerFilter(layerCand) {
          return layerCand === _this._layer;
        }
      }) || [];

      if ((0, _isFunction["default"])(onMapSingleClick)) {
        onMapSingleClick(olEvt, selectedFeatures);
      }

      selectedFeatures.forEach(function (selectedFeature) {
        var key = _this.props.keyFunction(selectedFeature);

        if (selectedRowKeys && selectedRowKeys.includes(key)) {
          selectedFeature.setStyle(null);

          var node = _this.getRowFromFeatureKey(key);

          if (node) {
            node.setSelected(false);
          }
        } else {
          selectedFeature.setStyle(selectStyle);

          var _node = _this.getRowFromFeatureKey(key);

          if (_node) {
            _node.setSelected(true);
          }
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "deinitVectorLayer", function () {
      var map = _this.props.map;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      map.removeLayer(_this._layer);
    });

    _defineProperty(_assertThisInitialized(_this), "deinitMapEventHandlers", function () {
      var _this$props4 = _this.props,
          map = _this$props4.map,
          selectable = _this$props4.selectable;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      map.un('pointermove', _this.onMapPointerMove);

      if (selectable) {
        map.un('singleclick', _this.onMapSingleClick);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getColumnDefs", function () {
      var _this$props5 = _this.props,
          attributeBlacklist = _this$props5.attributeBlacklist,
          features = _this$props5.features,
          columnDefs = _this$props5.columnDefs,
          selectable = _this$props5.selectable;
      var columns = [];
      var feature = features[0];

      if (!(feature instanceof _Feature["default"])) {
        return columns;
      }

      var props = feature.getProperties();

      if (selectable) {
        columns.push({
          headerName: '',
          checkboxSelection: true,
          headerCheckboxSelection: true,
          width: 28,
          pinned: 'left',
          lockPosition: true,
          suppressMenu: true,
          suppressSorting: true,
          suppressFilter: true,
          suppressResize: true,
          suppressMovable: true
        });
      }

      var index = -1;
      Object.keys(props).forEach(function (key) {
        if (attributeBlacklist.includes(key)) {
          return;
        }

        if (props[key] instanceof _Geometry["default"]) {
          return;
        }

        if (columnDefs[key] && columnDefs[key].index !== undefined) {
          index = columnDefs[key].index;
        } else {
          ++index;
        }

        columns[index] = _objectSpread({
          headerName: key,
          field: key
        }, columnDefs[key]);
      });
      return columns;
    });

    _defineProperty(_assertThisInitialized(_this), "getRowData", function () {
      var features = _this.props.features;
      var data = [];
      features.forEach(function (feature) {
        var properties = feature.getProperties();
        var filtered = Object.keys(properties).filter(function (key) {
          return !(properties[key] instanceof _Geometry["default"]);
        }).reduce(function (obj, key) {
          obj[key] = properties[key];
          return obj;
        }, {});
        data.push(_objectSpread({
          key: _this.props.keyFunction(feature)
        }, filtered));
      });
      return data;
    });

    _defineProperty(_assertThisInitialized(_this), "getFeatureFromRowKey", function (key) {
      var _this$props6 = _this.props,
          features = _this$props6.features,
          keyFunction = _this$props6.keyFunction;
      var feature = features.filter(function (feature) {
        return keyFunction(feature) === key;
      });
      return feature[0];
    });

    _defineProperty(_assertThisInitialized(_this), "getRowFromFeatureKey", function (key) {
      var grid = _this.state.grid;
      var rowNode;

      if (!grid || !grid.api) {
        return;
      }

      grid.api.forEachNode(function (node) {
        if (node.data.key === key) {
          rowNode = node;
        }
      });
      return rowNode;
    });

    _defineProperty(_assertThisInitialized(_this), "getSelectedRowKeys", function () {
      var grid = _this.state.grid;

      if (!grid || !grid.api) {
        return;
      }

      var selectedRows = grid.api.getSelectedRows();
      return selectedRows.map(function (row) {
        return row.key;
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onRowClick", function (evt) {
      var onRowClick = _this.props.onRowClick;
      var row = evt.data;

      var feature = _this.getFeatureFromRowKey(row.key);

      if ((0, _isFunction["default"])(onRowClick)) {
        onRowClick(row, feature, evt);
      } else {
        _this.zoomToFeatures([feature]);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onRowMouseOver", function (evt) {
      var onRowMouseOver = _this.props.onRowMouseOver;
      var row = evt.data;

      var feature = _this.getFeatureFromRowKey(row.key);

      if ((0, _isFunction["default"])(onRowMouseOver)) {
        onRowMouseOver(row, feature, evt);
      }

      _this.highlightFeatures([feature]);
    });

    _defineProperty(_assertThisInitialized(_this), "onRowMouseOut", function (evt) {
      var onRowMouseOut = _this.props.onRowMouseOut;
      var row = evt.data;

      var feature = _this.getFeatureFromRowKey(row.key);

      if ((0, _isFunction["default"])(onRowMouseOut)) {
        onRowMouseOut(row, feature, evt);
      }

      _this.unhighlightFeatures([feature]);
    });

    _defineProperty(_assertThisInitialized(_this), "zoomToFeatures", function (features) {
      var map = _this.props.map;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      var featGeometries = [];
      features.forEach(function (feature) {
        featGeometries.push(feature.getGeometry());
      });

      if (featGeometries.length > 0) {
        var geomCollection = new _GeometryCollection["default"](featGeometries);
        map.getView().fit(geomCollection.getExtent());
      }
    });

    _defineProperty(_assertThisInitialized(_this), "highlightFeatures", function (highlightFeatures) {
      var _this$props7 = _this.props,
          map = _this$props7.map,
          highlightStyle = _this$props7.highlightStyle;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      highlightFeatures.forEach(function (feature) {
        return feature.setStyle(highlightStyle);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "unhighlightFeatures", function (unhighlightFeatures) {
      var _this$props8 = _this.props,
          map = _this$props8.map,
          selectStyle = _this$props8.selectStyle;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      var selectedRowKeys = _this.getSelectedRowKeys();

      unhighlightFeatures.forEach(function (feature) {
        var key = _this.props.keyFunction(feature);

        if (selectedRowKeys && selectedRowKeys.includes(key)) {
          feature.setStyle(selectStyle);
        } else {
          feature.setStyle(null);
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "selectFeatures", function (features) {
      var _this$props9 = _this.props,
          map = _this$props9.map,
          selectStyle = _this$props9.selectStyle;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      features.forEach(function (feature) {
        if (feature) {
          feature.setStyle(selectStyle);
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "resetFeatureStyles", function () {
      var _this$props10 = _this.props,
          map = _this$props10.map,
          features = _this$props10.features;

      if (!(map instanceof _Map["default"])) {
        return;
      }

      features.forEach(function (feature) {
        return feature.setStyle(null);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onSelectionChanged", function (evt) {
      var onRowSelectionChange = _this.props.onRowSelectionChange;
      var _this$state = _this.state,
          grid = _this$state.grid,
          selectedRows = _this$state.selectedRows;
      var selectedRowsAfter;

      if (!grid || !grid.api) {
        selectedRowsAfter = evt.api.getSelectedRows();
      } else {
        selectedRowsAfter = grid.api.getSelectedRows();
      }

      var deselectedRows = (0, _differenceWith["default"])(selectedRows, selectedRowsAfter, function (a, b) {
        return a.key === b.key;
      });
      var selectedFeatures = selectedRowsAfter.map(function (row) {
        return _this.getFeatureFromRowKey(row.key);
      });
      var deselectedFeatures = deselectedRows.map(function (row) {
        return _this.getFeatureFromRowKey(row.key);
      }); // update state

      _this.setState({
        selectedRows: selectedRowsAfter
      });

      if ((0, _isFunction["default"])(onRowSelectionChange)) {
        onRowSelectionChange(selectedRowsAfter, selectedFeatures, deselectedRows, deselectedFeatures, evt);
      }

      _this.resetFeatureStyles();

      _this.selectFeatures(selectedFeatures);
    });

    _this.state = {
      grid: null,
      selectedRows: []
    };
    return _this;
  }
  /**
   * Called on lifecycle phase componentDidMount.
   */


  _createClass(AgFeatureGrid, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props11 = this.props,
          map = _this$props11.map,
          features = _this$props11.features,
          zoomToExtent = _this$props11.zoomToExtent;
      this.initVectorLayer(map);
      this.initMapEventHandlers(map);

      if (zoomToExtent) {
        this.zoomToFeatures(features);
      }
    }
    /**
     * Invoked immediately after updating occurs. This method is not called for
     * the initial render.
     *
     * @param {Object} prevProps The previous props.
     */

  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var _this$props12 = this.props,
          map = _this$props12.map,
          features = _this$props12.features,
          selectable = _this$props12.selectable,
          zoomToExtent = _this$props12.zoomToExtent;

      if (!(0, _isEqual["default"])(prevProps.map, map)) {
        this.initVectorLayer(map);
        this.initMapEventHandlers(map);
      }

      if (!(0, _isEqual["default"])(prevProps.features, features)) {
        if (this._source) {
          this._source.clear();

          this._source.addFeatures(features);
        }

        if (zoomToExtent) {
          this.zoomToFeatures(features);
        }
      }

      if (!(0, _isEqual["default"])(prevProps.selectable, selectable)) {
        if (selectable && map) {
          map.on('singleclick', this.onMapSingleClick);
        } else {
          if (map) {
            map.un('singleclick', this.onMapSingleClick);
          }
        }
      }
    }
    /**
     * Called on lifecycle phase componentWillUnmount.
     */

  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.deinitVectorLayer();
      this.deinitMapEventHandlers();
    }
    /**
     * Initialized the vector layer that will be used to draw the input features
     * on and adds it to the map (if any).
     *
     * @param {ol.Map} map The map to add the layer to.
     */

  }, {
    key: "onGridReady",

    /**
     *
     * @param {*} grid
     */
    value: function onGridReady(grid) {
      this.setState({
        grid: grid
      }, this.onVisiblityChange);

      if (this.props.onGridIsReady) {
        this.props.onGridIsReady(grid);
      }
    }
    /**
     *
     */

  }, {
    key: "onVisiblityChange",
    value: function onVisiblityChange() {
      if (this.state.grid) {
        this.state.grid.api.sizeColumnsToFit();
      }
    }
    /**
     * The render method.
     */

  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props13 = this.props,
          className = _this$props13.className,
          height = _this$props13.height,
          width = _this$props13.width,
          theme = _this$props13.theme,
          rowClassName = _this$props13.rowClassName,
          features = _this$props13.features,
          map = _this$props13.map,
          attributeBlacklist = _this$props13.attributeBlacklist,
          onRowClick = _this$props13.onRowClick,
          onRowMouseOver = _this$props13.onRowMouseOver,
          onRowMouseOut = _this$props13.onRowMouseOut,
          zoomToExtent = _this$props13.zoomToExtent,
          selectable = _this$props13.selectable,
          featureStyle = _this$props13.featureStyle,
          highlightStyle = _this$props13.highlightStyle,
          selectStyle = _this$props13.selectStyle,
          layerName = _this$props13.layerName,
          columnDefs = _this$props13.columnDefs,
          children = _this$props13.children,
          rowData = _this$props13.rowData,
          passThroughProps = _objectWithoutProperties(_this$props13, ["className", "height", "width", "theme", "rowClassName", "features", "map", "attributeBlacklist", "onRowClick", "onRowMouseOver", "onRowMouseOut", "zoomToExtent", "selectable", "featureStyle", "highlightStyle", "selectStyle", "layerName", "columnDefs", "children", "rowData"]);

      var finalClassName = className ? "".concat(className, " ").concat(this._className, " ").concat(theme) : "".concat(this._className, " ").concat(theme); // TODO: Not sure, if this is still needed. One may want to get a specific
      // row by using getRowFromFeatureKey instead.

      var rowClassNameFn;

      if ((0, _isFunction["default"])(rowClassName)) {
        rowClassNameFn = function rowClassNameFn(node) {
          var determinedRowClass = rowClassName(node.data);
          return "".concat(_this2._rowClassName, " ").concat(determinedRowClass);
        };
      } else {
        var finalRowClassName = rowClassName ? "".concat(rowClassName, " ").concat(this._rowClassName) : this._rowClassName;

        rowClassNameFn = function rowClassNameFn(node) {
          return "".concat(finalRowClassName, " ").concat(_this2._rowKeyClassNamePrefix).concat((0, _kebabCase["default"])(node.data.key));
        };
      }

      return _react["default"].createElement("div", {
        className: finalClassName,
        style: {
          height: height,
          width: width
        }
      }, _react["default"].createElement(_agGridReact.AgGridReact, _extends({
        columnDefs: columnDefs && (0, _isArray["default"])(columnDefs) ? columnDefs : this.getColumnDefs(),
        rowData: rowData && (0, _isArray["default"])(rowData) ? rowData : this.getRowData(),
        onGridReady: this.onGridReady.bind(this),
        rowSelection: "multiple",
        suppressRowClickSelection: true,
        onSelectionChanged: this.onSelectionChanged.bind(this),
        onRowClicked: this.onRowClick.bind(this),
        onCellMouseOver: this.onRowMouseOver.bind(this),
        onCellMouseOut: this.onRowMouseOut.bind(this),
        ref: function ref(_ref) {
          return _this2._ref = _ref;
        },
        getRowClass: rowClassNameFn
      }, passThroughProps), children));
    }
  }]);

  return AgFeatureGrid;
}(_react["default"].Component);

exports.AgFeatureGrid = AgFeatureGrid;

_defineProperty(AgFeatureGrid, "propTypes", {
  /**
   * An optional CSS class which should be added to the table.
   * @type {String}
   */
  className: _propTypes["default"].string,

  /**
   * The height of the grid.
   *
   * @type {Number|String}
   */
  height: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),

  /**
   * The width of the grid.
   *
   * @type {Number|String}
   */
  width: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),

  /**
   * The theme to use for the grid. See
   * https://www.ag-grid.com/javascript-grid-styling/ for available options.
   * Note: CSS must be loaded to use the theme!
   */
  theme: _propTypes["default"].string,

  /**
   * An optional CSS class to add to each table row or a function that
   * is evaluated for each record.
   * @type {String|Function}
   */
  rowClassName: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func]),

  /**
   * The features to show in the grid and the map (if set).
   * @type {Array}
   */
  features: _propTypes["default"].arrayOf(_propTypes["default"].instanceOf(_Feature["default"])),

  /**
   * The map the features should be rendered on. If not given, the features
   * will be rendered in the table only.
   * @type {ol.Map}
   */
  map: _propTypes["default"].instanceOf(_Map["default"]),

  /**
   * A list of attribute names to hide in the table.
   * @type {Array}
   */
  attributeBlacklist: _propTypes["default"].arrayOf(_propTypes["default"].string),

  /**
   * Optional callback function, that will be called on rowclick.
   * @type {Function}
   */
  onRowClick: _propTypes["default"].func,

  /**
   * Optional callback function, that will be called on rowmouseover.
   * @type {Function}
   */
  onRowMouseOver: _propTypes["default"].func,

  /**
   * Optional callback function, that will be called on rowmouseout.
   * @type {Function}
   */
  onRowMouseOut: _propTypes["default"].func,

  /**
   * Optional callback function, that will be called if the selection changes.
   * @type {Function}
   */
  onRowSelectionChange: _propTypes["default"].func,

  /**
   * Optional callback function, that will be called if `selectable` is set
   * `true` and the a `click` event on the map occurs, e.g. a feature has been
   * selected in the map. The function receives the olEvt and the selected
   * features (if any).
   * @type {Function}
   */
  onMapSingleClick: _propTypes["default"].func,

  /**
   * Whether the map should center on the current feature's extent on init or
   * not.
   * @type {Boolean}
   */
  zoomToExtent: _propTypes["default"].bool,

  /**
   * Whether features should be selectable via a map click or not. If you want
   * to enable/disable a checkbox, please set checkboxSelection on a column
   * in columnDefs.
   *
   * @type {Boolean}
   */
  selectable: _propTypes["default"].bool,

  /**
   * The default style to apply to the features.
   * @type {ol.Style|ol.FeatureStyleFunction}
   */
  featureStyle: _propTypes["default"].oneOfType([_propTypes["default"].instanceOf(_Style["default"]), _propTypes["default"].func]),

  /**
   * The highlight style to apply to the features.
   * @type {ol.Style|ol.FeatureStyleFunction}
   */
  highlightStyle: _propTypes["default"].oneOfType([_propTypes["default"].instanceOf(_Style["default"]), _propTypes["default"].func]),

  /**
   * The select style to apply to the features.
   * @type {ol.Style|ol.FeatureStyleFunction}
   */
  selectStyle: _propTypes["default"].oneOfType([_propTypes["default"].instanceOf(_Style["default"]), _propTypes["default"].func]),

  /**
   * The name of the vector layer presenting the features in the grid.
   * @type {String}
   */
  layerName: _propTypes["default"].string,

  /**
   * Custom column definitions to apply to the given column (mapping via key).
   * See https://ant.design/components/table/#Column. You can either specify
   * an index property on every column definition to get an exact order, or
   * get a somewhat random order by not specifying an index property at all.
   * If provided as array, #getColumnDefs won't be called.
   * @type {Object|Array}
   */
  columnDefs: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].arrayOf(_propTypes["default"].object)]),

  /**
   * Custom row data to be shown in feature grid. This might be helpful if
   * original feature properties should be manipulated in some way before they
   * are represented in grid.
   * If provided, #getRowData method won't be called.
   * @type {Array}
   */
  rowData: _propTypes["default"].arrayOf(_propTypes["default"].object),

  /**
   * The children to render.
   * @type {Element}
   */
  children: _propTypes["default"].element,

  /**
   * A Function that creates the rowkey from the given feature.
   * Receives the feature as property.
   * Default is: feature => feature.ol_uid
   *
   * @type {Function}
   */
  keyFunction: _propTypes["default"].func,

  /**
   * A Function that is called once the grid is ready.
   * @type {Function}
   */
  onGridIsReady: _propTypes["default"].func
});

_defineProperty(AgFeatureGrid, "defaultProps", {
  theme: 'ag-theme-balham',
  height: 250,
  features: [],
  attributeBlacklist: [],
  featureStyle: new _Style["default"]({
    fill: new _Fill["default"]({
      color: 'rgba(255, 255, 255, 0.5)'
    }),
    stroke: new _Stroke["default"]({
      color: 'rgba(73, 139, 170, 0.9)',
      width: 1
    }),
    image: new _Circle["default"]({
      radius: 6,
      fill: new _Fill["default"]({
        color: 'rgba(255, 255, 255, 0.5)'
      }),
      stroke: new _Stroke["default"]({
        color: 'rgba(73, 139, 170, 0.9)',
        width: 1
      })
    })
  }),
  highlightStyle: new _Style["default"]({
    fill: new _Fill["default"]({
      color: 'rgba(230, 247, 255, 0.8)'
    }),
    stroke: new _Stroke["default"]({
      color: 'rgba(73, 139, 170, 0.9)',
      width: 1
    }),
    image: new _Circle["default"]({
      radius: 6,
      fill: new _Fill["default"]({
        color: 'rgba(230, 247, 255, 0.8)'
      }),
      stroke: new _Stroke["default"]({
        color: 'rgba(73, 139, 170, 0.9)',
        width: 1
      })
    })
  }),
  selectStyle: new _Style["default"]({
    fill: new _Fill["default"]({
      color: 'rgba(230, 247, 255, 0.8)'
    }),
    stroke: new _Stroke["default"]({
      color: 'rgba(73, 139, 170, 0.9)',
      width: 2
    }),
    image: new _Circle["default"]({
      radius: 6,
      fill: new _Fill["default"]({
        color: 'rgba(230, 247, 255, 0.8)'
      }),
      stroke: new _Stroke["default"]({
        color: 'rgba(73, 139, 170, 0.9)',
        width: 2
      })
    })
  }),
  layerName: 'react-geo-feature-grid-layer',
  columnDefs: {},
  keyFunction: function keyFunction(feature) {
    return feature.ol_uid;
  }
  /**
   * The constructor.
   */

});

var _default = AgFeatureGrid;
exports["default"] = _default;
