// @terrestris/ol-util/dist/CapabilitiesUtil/CapabilitiesUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _WMSCapabilities = _interopRequireDefault(require("ol/format/WMSCapabilities"));

var _ImageWMS = _interopRequireDefault(require("ol/source/ImageWMS"));

var _Image = _interopRequireDefault(require("ol/layer/Image"));

var _get = _interopRequireDefault(require("lodash/get.js"));

var _isFunction = _interopRequireDefault(require("lodash/isFunction"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Helper class to parse capabilities of WMS layers
 *
 * @class CapabilitiesUtil
 */
var CapabilitiesUtil =
/*#__PURE__*/
function () {
  function CapabilitiesUtil() {
    _classCallCheck(this, CapabilitiesUtil);
  }

  _createClass(CapabilitiesUtil, null, [{
    key: "parseWmsCapabilities",

    /**
     * Parses the given WMS Capabilities string.
     *
     * @param {string} capabilitiesUrl Url to WMS capabilities document
     * @return {Object} An object representing the WMS capabilities.
     */
    value: function parseWmsCapabilities(capabilitiesUrl) {
      return fetch(capabilitiesUrl).then(function (response) {
        return response.text();
      }).then(function (data) {
        var wmsCapabilitiesParser = new _WMSCapabilities["default"]();
        return wmsCapabilitiesParser.read(data);
      });
    }
    /**
     * Returns the layers from a parsed WMS GetCapabilities object.
     *
     * @param {Object} capabilities A capabilities object.
     * @param {string} nameField Configure the field which should be set as the
     *                           'name' property in the openlayers layer.
     * @param {Function} proxyFn Optional proxy function which can be applied to
     *                           `GetMap`, `GetFeatureInfo` and `GetLegendGraphic`
     *                           requests to avoid CORS issues.
     * @return {OlLayerTile[]} Array of OlLayerTile
     */

  }, {
    key: "getLayersFromWmsCapabilities",
    value: function getLayersFromWmsCapabilities(capabilities) {
      var nameField = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'Name';
      var proxyFn = arguments.length > 2 ? arguments[2] : undefined;
      var wmsVersion = (0, _get["default"])(capabilities, 'version');
      var wmsAttribution = (0, _get["default"])(capabilities, 'Service.AccessConstraints');
      var layersInCapabilities = (0, _get["default"])(capabilities, 'Capability.Layer.Layer');
      var wmsGetMapConfig = (0, _get["default"])(capabilities, 'Capability.Request.GetMap');
      var wmsGetFeatureInfoConfig = (0, _get["default"])(capabilities, 'Capability.Request.GetFeatureInfo');
      var getMapUrl = (0, _get["default"])(wmsGetMapConfig, 'DCPType[0].HTTP.Get.OnlineResource');
      var getFeatureInfoUrl = (0, _get["default"])(wmsGetFeatureInfoConfig, 'DCPType[0].HTTP.Get.OnlineResource');
      var legendUrl = layersInCapabilities.length > 0 ? (0, _get["default"])(layersInCapabilities[0], 'Style[0].LegendURL[0].OnlineResource') : null;
      return layersInCapabilities.map(function (layerObj) {
        return new _Image["default"]({
          opacity: 1,
          title: (0, _get["default"])(layerObj, 'Title'),
          name: (0, _get["default"])(layerObj, nameField),
          "abstract": (0, _get["default"])(layerObj, 'Abstract'),
          getFeatureInfoUrl: (0, _isFunction["default"])(proxyFn) ? proxyFn(getFeatureInfoUrl) : getFeatureInfoUrl,
          getFeatureInfoFormats: (0, _get["default"])(wmsGetFeatureInfoConfig, 'Format'),
          legendUrl: (0, _isFunction["default"])(proxyFn) ? proxyFn(legendUrl) : legendUrl,
          queryable: (0, _get["default"])(layerObj, 'queryable'),
          source: new _ImageWMS["default"]({
            url: (0, _isFunction["default"])(proxyFn) ? proxyFn(getMapUrl) : getMapUrl,
            attributions: wmsAttribution,
            params: {
              'LAYERS': (0, _get["default"])(layerObj, 'Name'),
              'VERSION': wmsVersion
            }
          })
        });
      });
    }
  }]);

  return CapabilitiesUtil;
}();

var _default = CapabilitiesUtil;
exports["default"] = _default;
