// @terrestris/react-geo/dist/Field/WfsSearchInput/WfsSearchInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.WfsSearchInput = void 0;

require("antd/es/input/style");

var _input = _interopRequireDefault(require("antd/es/input"));

require("antd/es/icon/style");

var _icon = _interopRequireDefault(require("antd/es/icon"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Map = _interopRequireDefault(require("ol/Map"));

var _isFunction = _interopRequireDefault(require("lodash/isFunction"));

var _debounce = _interopRequireDefault(require("lodash/debounce"));

var _Logger = _interopRequireDefault(require("@terrestris/base-util/dist/Logger"));

var _WfsFilterUtil = _interopRequireDefault(require("@terrestris/ol-util/dist/WfsFilterUtil/WfsFilterUtil"));

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { if (i % 2) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } else { Object.defineProperties(target, Object.getOwnPropertyDescriptors(arguments[i])); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The WfsSearchInput field.
 * Implements an input field to do a WFS-GetFeature request.
 *
 * The GetFeature request is created with `ol.format.WFS.writeGetFeature`
 * so most of the WFS specific options work like document in the corresponding
 * API-docs: https://openlayers.org/en/latest/apidoc/module-ol_format_WFS.html
 *
 * @class WfsSearchInput
 * @extends React.Component
 */
var WfsSearchInput =
/*#__PURE__*/
function (_React$Component) {
  _inherits(WfsSearchInput, _React$Component);

  /**
   * The className added to this component.
   * @type {String}
   * @private
   */
  function WfsSearchInput(props) {
    var _this;

    _classCallCheck(this, WfsSearchInput);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(WfsSearchInput).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "wfssearchinput"));

    _this.state = {
      searchTerm: '',
      data: []
    };
    _this.onUpdateInput = _this.onUpdateInput.bind(_assertThisInitialized(_this)); // delay requests invoking

    _this.doSearch = (0, _debounce["default"])(_this.doSearch, _this.props.delay);
    return _this;
  }
  /**
   * Called if the input is being updated. It sets the current inputValue
   * as searchTerm and starts a search if the inputValue has
   * a length of at least `this.props.minChars` (default 3).
   *
   * @param {Event|undefined} evt The input value from `keyup` event.
   * Gets undefined if clear btn is pressed.
   */


  _createClass(WfsSearchInput, [{
    key: "onUpdateInput",
    value: function onUpdateInput(evt) {
      var _this2 = this;

      var _this$props = this.props,
          minChars = _this$props.minChars,
          onBeforeSearch = _this$props.onBeforeSearch;
      this.setState({
        data: []
      });
      var value = '';

      if (evt && evt.target && evt.target.value) {
        value = evt.target.value;
      }

      if ((0, _isFunction["default"])(onBeforeSearch)) {
        value = onBeforeSearch(value);
      }

      this.setState({
        searchTerm: value
      }, function () {
        if (_this2.state.searchTerm && _this2.state.searchTerm.length >= minChars) {
          _this2.doSearch();
        }
      });
    }
    /**
     * Perform the search.
     * @private
     */

  }, {
    key: "doSearch",
    value: function doSearch() {
      var _this3 = this;

      var _this$props2 = this.props,
          additionalFetchOptions = _this$props2.additionalFetchOptions,
          baseUrl = _this$props2.baseUrl,
          featureNS = _this$props2.featureNS,
          featurePrefix = _this$props2.featurePrefix,
          featureTypes = _this$props2.featureTypes,
          geometryName = _this$props2.geometryName,
          maxFeatures = _this$props2.maxFeatures,
          outputFormat = _this$props2.outputFormat,
          propertyNames = _this$props2.propertyNames,
          srsName = _this$props2.srsName,
          wfsFormatOptions = _this$props2.wfsFormatOptions,
          searchAttributes = _this$props2.searchAttributes,
          attributeDetails = _this$props2.attributeDetails;
      var searchOpts = {
        featureNS: featureNS,
        featurePrefix: featurePrefix,
        featureTypes: featureTypes,
        geometryName: geometryName,
        maxFeatures: maxFeatures,
        outputFormat: outputFormat,
        propertyNames: propertyNames,
        srsName: srsName,
        wfsFormatOptions: wfsFormatOptions,
        searchAttributes: searchAttributes,
        attributeDetails: attributeDetails
      };

      var request = _WfsFilterUtil["default"].getCombinedRequests(searchOpts, this.state.searchTerm);

      if (request) {
        this.setState({
          fetching: true
        }, function () {
          fetch("".concat(baseUrl), _objectSpread({
            method: 'POST',
            credentials: additionalFetchOptions.credentials ? additionalFetchOptions.credentials : 'same-origin',
            body: new XMLSerializer().serializeToString(request)
          }, additionalFetchOptions)).then(function (response) {
            return response.json();
          }).then(_this3.onFetchSuccess.bind(_this3))["catch"](_this3.onFetchError.bind(_this3));
        });
      } else {
        this.onFetchError('Missing GetFeature request parameters');
      }
    }
    /**
     * This function gets called on success of the WFS GetFeature fetch request.
     * It sets the response as data.
     * If an additional function `onFetchSuccess` is provided, it will be called
     * as callback.
     * @param {Array<object>} response The found features.
     */

  }, {
    key: "onFetchSuccess",
    value: function onFetchSuccess(response) {
      var _this4 = this;

      var onFetchSuccess = this.props.onFetchSuccess;
      var data = response.features ? response.features : [];
      data.forEach(function (feature) {
        return feature.searchTerm = _this4.state.searchTerm;
      });
      this.setState({
        data: data,
        fetching: false
      }, function () {
        if ((0, _isFunction["default"])(onFetchSuccess)) {
          onFetchSuccess(data);
        }
      });
    }
    /**
     * This function gets called when the WFS GetFeature fetch request returns an error.
     * It logs the error to the console.
     * If an additional function `onFetchSuccess` is provided, it will be called
     * as callback.
     *
     * @param {String} error The errorstring.
     */

  }, {
    key: "onFetchError",
    value: function onFetchError(error) {
      var onFetchError = this.props.onFetchError;

      _Logger["default"].error("Error while requesting WFS GetFeature: ".concat(error));

      this.setState({
        fetching: false
      }, function () {
        if ((0, _isFunction["default"])(onFetchError)) {
          onFetchError(error);
        }
      });
    }
    /**
     * Resets input field value and previously fetched data on reset button click.
     */

  }, {
    key: "resetSearch",
    value: function resetSearch() {
      var onClearClick = this.props.onClearClick;
      this.inputRef.input.value = '';
      this.setState({
        data: []
      }, function () {
        if ((0, _isFunction["default"])(onClearClick)) {
          onClearClick();
        }
      });
    }
    /**
     * The render function.
     */

  }, {
    key: "render",
    value: function render() {
      var _this5 = this;

      var fetching = this.state.fetching;

      var _this$props3 = this.props,
          additionalFetchOptions = _this$props3.additionalFetchOptions,
          baseUrl = _this$props3.baseUrl,
          className = _this$props3.className,
          featureNS = _this$props3.featureNS,
          featurePrefix = _this$props3.featurePrefix,
          featureTypes = _this$props3.featureTypes,
          filter = _this$props3.filter,
          geometryName = _this$props3.geometryName,
          map = _this$props3.map,
          maxFeatures = _this$props3.maxFeatures,
          minChars = _this$props3.minChars,
          outputFormat = _this$props3.outputFormat,
          propertyNames = _this$props3.propertyNames,
          searchAttributes = _this$props3.searchAttributes,
          attributeDetails = _this$props3.attributeDetails,
          srsName = _this$props3.srsName,
          wfsFormatOptions = _this$props3.wfsFormatOptions,
          onFetchSuccess = _this$props3.onFetchSuccess,
          onFetchError = _this$props3.onFetchError,
          onClearClick = _this$props3.onClearClick,
          onBeforeSearch = _this$props3.onBeforeSearch,
          passThroughProps = _objectWithoutProperties(_this$props3, ["additionalFetchOptions", "baseUrl", "className", "featureNS", "featurePrefix", "featureTypes", "filter", "geometryName", "map", "maxFeatures", "minChars", "outputFormat", "propertyNames", "searchAttributes", "attributeDetails", "srsName", "wfsFormatOptions", "onFetchSuccess", "onFetchError", "onClearClick", "onBeforeSearch"]);

      var finalClassName = className ? "".concat(className, " ").concat(this.className) : this.className;
      return _react["default"].createElement(_input["default"], _extends({
        className: finalClassName,
        ref: function ref(inputRef) {
          _this5.inputRef = inputRef;
        },
        suffix: _react["default"].createElement(_icon["default"], {
          type: fetching ? 'loading' : 'close-circle',
          onClick: this.resetSearch.bind(this)
        }),
        onPressEnter: this.onUpdateInput,
        onKeyUp: this.onUpdateInput
      }, passThroughProps));
    }
  }]);

  return WfsSearchInput;
}(_react["default"].Component);

exports.WfsSearchInput = WfsSearchInput;

_defineProperty(WfsSearchInput, "propTypes", {
  /**
   * An optional CSS class which should be added.
   * @type {String}
   */
  className: _propTypes["default"].string,

  /**
   * The base URL. Please make sure that the WFS-Server supports CORS.
   * @type {String}
   */
  baseUrl: _propTypes["default"].string.isRequired,

  /**
   * An object mapping feature types to an array of attributes that should be searched through.
   * @type {Object}
   */
  searchAttributes: _propTypes["default"].object.isRequired,

  /**
   * A nested object mapping feature types to an object of attribute details,
   * which are also mapped by search attribute name.
   *
   * Example:
   *   ```
   *   attributeDetails: {
   *    featType1: {
   *      attr1: {
   *        matchCase: true,
   *        type: 'number',
   *        exactSearch: false
   *      },
   *      attr2: {
   *        matchCase: false,
   *        type: 'string',
   *        exactSearch: true
   *      }
   *    },
   *    featType2: {...}
   *   }
   *   ```
   * @type {Object}
   */
  attributeDetails: _propTypes["default"].objectOf(_propTypes["default"].objectOf(_propTypes["default"].shape({
    matchCase: _propTypes["default"].bool,
    type: _propTypes["default"].string,
    exactSearch: _propTypes["default"].bool
  }))),

  /**
   * The namespace URI used for features.
   * @type {String}
   */
  featureNS: _propTypes["default"].string,

  /**
   * The prefix for the feature namespace.
   * @type {String}
   */
  featurePrefix: _propTypes["default"].string,

  /**
   * The feature type names. Required.
   * @type {String[]}
   */
  featureTypes: _propTypes["default"].arrayOf(_propTypes["default"].string).isRequired,

  /**
   * SRS name. No srsName attribute will be set on geometries when this is not
   * provided.
   * @type {String}
   */
  srsName: _propTypes["default"].string,

  /**
   * The output format of the response.
   * @type {String}
   */
  outputFormat: _propTypes["default"].string,

  /**
   * Maximum number of features to fetch.
   * @type {Number}
   */
  maxFeatures: _propTypes["default"].number,

  /**
   * Geometry name to use in a BBOX filter.
   * @type {String}
   */
  geometryName: _propTypes["default"].string,

  /**
   * Optional list of property names to serialize.
   * @type {String[]}
   */
  propertyNames: _propTypes["default"].arrayOf(_propTypes["default"].string),

  /**
   * Filter condition. See https://openlayers.org/en/latest/apidoc/module-ol_format_filter.html.
   * for more information.
   * @type {Object}
   */
  filter: _propTypes["default"].object,

  /**
   * The ol.map to interact with on selection.
   *
   * @type {ol.Map}
   */
  map: _propTypes["default"].instanceOf(_Map["default"]),

  /**
   * The minimal amount of characters entered in the input to start a search.
   * @type {Number}
   */
  minChars: _propTypes["default"].number,

  /**
   * An onFetchSuccess callback function which gets called with the
   * successfully fetched data.
   * Please note: if omitted only data fetch will be performed and no data
   * will be shown afterwards!
   * @type {Function}
   */
  onFetchSuccess: _propTypes["default"].func,

  /**
   * An onFetchError callback function which gets called if data fetch is
   * failed.
   * @type {Function}
   */
  onFetchError: _propTypes["default"].func,

  /**
   * Optional callback function, that will be called if 'clear' button of
   * input field was clicked.
   * @type {Function}
   */
  onClearClick: _propTypes["default"].func,

  /**
    * Optional callback function, that will be called before WFS search starts.
    * Can be useful if input value manipulation is needed before (e.g. umlaut
    * replacement `ä => ae` etc.)
    * @type {Function}
    */
  onBeforeSearch: _propTypes["default"].func,

  /**
   * Options which are added to the fetch-POST-request. credentials is set to
   * 'same-origin' as default but can be overwritten. See also
   * https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
   * @type {Object}
   */
  additionalFetchOptions: _propTypes["default"].object,

  /**
   * Options which are passed to the constructor of the ol.format.WFS.
   * compare: https://openlayers.org/en/latest/apidoc/module-ol_format_WFS.html
   * @type {Object}
   */
  wfsFormatOptions: _propTypes["default"].object,

  /**
   * Delay in ms before actually sending requests.
   * @type {Number}
   */
  delay: _propTypes["default"].number
});

_defineProperty(WfsSearchInput, "defaultProps", {
  srsName: 'EPSG:3857',
  outputFormat: 'application/json',
  minChars: 3,
  additionalFetchOptions: {},
  attributeDetails: {},
  delay: 300
  /**
   * Create the WfsSearchInput.
   *
   * @param {Object} props The initial props.
   * @constructs WfsSearchInput
   */

});

var _default = WfsSearchInput;
exports["default"] = _default;
