// @terrestris/react-geo/dist/HigherOrderComponent/LoadifiedComponent/LoadifiedComponent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadify = loadify;

require("antd/es/spin/style");

var _spin = _interopRequireDefault(require("antd/es/spin"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Logger = _interopRequireDefault(require("@terrestris/base-util/dist/Logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The HOC factory function.
 *
 * Wrapped components will be used as the content of the Loader component.
 * If the wrapped component is set to be loading the loader element
 * will be shown and if not it wont.
 *
 * @param {Component} WrappedComponent The component to wrap and enhance.
 * @param {Component} options The options to apply.
 * @return {Component} The wrapped component.
 */
function loadify(WrappedComponent) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$withRef = _ref.withRef,
      withRef = _ref$withRef === void 0 ? false : _ref$withRef;

  /**
   * The wrapper class for the given component.
   *
   * @class The loadify
   * @extends React.Component
   */
  var LoadifiedComponent =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(LoadifiedComponent, _React$Component);

    /**
     * The props.
     * @type {Object}
     */
    function LoadifiedComponent(props) {
      var _this;

      _classCallCheck(this, LoadifiedComponent);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(LoadifiedComponent).call(this, props));
      /**
       * The wrapped instance.
       * @type {Element}
       */

      _defineProperty(_assertThisInitialized(_this), "getWrappedInstance", function () {
        if (withRef) {
          return _this.wrappedInstance;
        } else {
          _Logger["default"].debug('No wrapped instance referenced, please call the ' + 'Loadify with option withRef = true.');
        }
      });

      _defineProperty(_assertThisInitialized(_this), "setWrappedInstance", function (instance) {
        if (withRef) {
          _this.wrappedInstance = instance;
        }
      });

      _this.wrappedInstance = null;
      return _this;
    }
    /**
     * Returns the wrapped instance. Only applicable if withRef is set to true.
     *
     * @return {Element} The wrappend instance.
     */


    _createClass(LoadifiedComponent, [{
      key: "render",

      /**
       * The render function.
       */
      value: function render() {
        var _this$props = this.props,
            size = _this$props.size,
            indicator = _this$props.indicator,
            tip = _this$props.tip,
            delay = _this$props.delay,
            spinning = _this$props.spinning,
            passThroughProps = _objectWithoutProperties(_this$props, ["size", "indicator", "tip", "delay", "spinning"]); // Propoerties passed to the Spin component. For the List of all the Spin component
        // properties see:
        // https://ant.design/components/spin/


        var passToLoader = {
          size: size,
          indicator: indicator,
          tip: tip,
          delay: delay,
          spinning: spinning
        };
        return _react["default"].createElement(_spin["default"], passToLoader, _react["default"].createElement(WrappedComponent, _extends({
          ref: this.setWrappedInstance
        }, passThroughProps)));
      }
    }]);

    return LoadifiedComponent;
  }(_react["default"].Component);

  _defineProperty(LoadifiedComponent, "propTypes", {
    /**
     * Whether the component should be loading or not.
     * @type {Boolean}
     */
    spinning: _propTypes["default"].bool,

    /**
     * Size of the loading element.
     * @type {string}
     */
    size: _propTypes["default"].oneOf(['small', 'default', 'large']),

    /**
     * The indicator element to use for the loader.
     * @type {String}
     */
    indicator: _propTypes["default"].any,

    /**
     * The tip text of loader element
     * @type {String}
     */
    tip: _propTypes["default"].string,

    /**
     * The delay time for the loader to show.
     * @type {Number}
     */
    delay: _propTypes["default"].number
    /**
     * The default properties.
     *
     * @type {Object}
     */

  });

  _defineProperty(LoadifiedComponent, "defaultProps", {
    spinning: false
    /**
     * Create the Loadify.
     *
     * @constructs Loadify
     */

  });

  return LoadifiedComponent;
}
