// @terrestris/react-geo/dist/Button/ToggleGroup/ToggleGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _isFunction = _interopRequireDefault(require("lodash/isFunction"));

var _constants = require("../../constants");

require("./ToggleGroup.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * A group for toggle components (e.g. buttons)
 *
 * @class The ToggleGroup
 * @extends React.Component
 */
var ToggleGroup =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ToggleGroup, _React$Component);

  /**
   * The className added to this component.
   * @type {String}
   * @private
   */

  /**
   * The properties.
   * @type {Object}
   */
  function ToggleGroup(props) {
    var _this;

    _classCallCheck(this, ToggleGroup);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ToggleGroup).call(this, props));
    /**
     * The initial state.
     * @type {Object}
     */

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "togglegroup"));

    _defineProperty(_assertThisInitialized(_this), "onChange", function (childProps) {
      if ((0, _isFunction["default"])(_this.props.onChange)) {
        _this.props.onChange(childProps);
      } // Allow deselect.


      if (_this.props.allowDeselect && childProps.name === _this.state.selectedName) {
        _this.setState({
          selectedName: null
        });
      } else {
        _this.setState({
          selectedName: childProps.name
        });
      }
    });

    _this.state = {
      selectedName: props.selectedName
    };
    return _this;
  }
  /**
   * Returns the context for the children.
   *
   * @return {Object} The child context.
   */


  _createClass(ToggleGroup, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        toggleGroup: {
          name: this.props.name,
          selectedName: this.state.selectedName,
          onChange: this.onChange
        }
      };
    }
    /**
     * The onChange handler.
     *
     * @param {Object} childProps The properties if the children.
     */

  }, {
    key: "render",

    /**
     * The render function.
     */
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          orientation = _this$props.orientation,
          children = _this$props.children;
      var className = this.props.className ? "".concat(this.props.className, " ").concat(this.className) : this.className;
      var orientationClass = orientation === 'vertical' ? 'vertical-toggle-group' : 'horizontal-toggle-group';

      var childrenWithProps = _react["default"].Children.map(children, function (child) {
        // pass the press state through to child components
        return _react["default"].cloneElement(child, {
          pressed: _this2.state.selectedName === child.props.name
        });
      });

      return _react["default"].createElement("div", {
        className: "".concat(className, " ").concat(orientationClass)
      }, childrenWithProps);
    }
  }]);

  return ToggleGroup;
}(_react["default"].Component);

_defineProperty(ToggleGroup, "propTypes", {
  /**
   * The className which should be added.
   * @type {String}
   */
  className: _propTypes["default"].string,

  /**
   * The name of this group.
   * @type {String}
   */
  name: _propTypes["default"].string,

  /**
   * The orientation of the children. Either `vertical` (default)
   * or `horizontal`.
   * @type {String}
   */
  orientation: _propTypes["default"].string,

  /**
   * Whether it's allowed to deselect a children or not.
   * @type {Boolean}
   */
  allowDeselect: _propTypes["default"].bool,

  /**
   * The value fo the `name` attribute of the children to select/press
   * initially.
   * @type {String}
   */
  selectedName: _propTypes["default"].string,

  /**
   * Callback function for onChange.
   * @type {Function}
   */
  onChange: _propTypes["default"].func,

  /**
   * The children of this group. Typically a set of `ToggleButton`s.
   * @type {Object}
   */
  children: _propTypes["default"].node
  /**
   * The default properties.
   * @type {Object}
   */

});

_defineProperty(ToggleGroup, "defaultProps", {
  orientation: 'vertical',
  allowDeselect: true
  /**
   * The child context types.
   * @type {Object}
   */

});

_defineProperty(ToggleGroup, "childContextTypes", {
  toggleGroup: _propTypes["default"].object
  /**
   * The child context types.
   * @type {Object}
   */

});

_defineProperty(ToggleGroup, "childContextTypes", {
  toggleGroup: _propTypes["default"].object
  /**
   * The constructor.
   *
   * @param {Object} props The properties.
   */

});

var _default = ToggleGroup;
exports["default"] = _default;
