// @terrestris/react-geo/dist/CircleMenu/CircleMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.CircleMenu = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _CircleMenuItem = _interopRequireDefault(require("./CircleMenuItem/CircleMenuItem.js"));

var _constants = require("../constants");

require("./CircleMenu.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { if (i % 2) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } else { Object.defineProperties(target, Object.getOwnPropertyDescriptors(arguments[i])); } } return target; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The CircleMenu.
 *
 * @class CircleMenu
 * @extends React.Component
 */
var CircleMenu =
/*#__PURE__*/
function (_React$Component) {
  _inherits(CircleMenu, _React$Component);

  function CircleMenu() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, CircleMenu);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(CircleMenu)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "circlemenu"));

    _defineProperty(_assertThisInitialized(_this), "_ref", null);

    return _this;
  }

  _createClass(CircleMenu, [{
    key: "componentDidMount",

    /**
     * A react lifecycle method called when the component did mount.
     * It calls the applyTransformation method right after mounting.
     */
    value: function componentDidMount() {
      setTimeout(this.applyTransformation.bind(this), 1);
    }
    /**
     * A react lifecycle method called when the component did update.
     * It calls the applyTransformation method right after updating.
     */

  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      setTimeout(this.applyTransformation.bind(this), 1);
    }
    /**
     * Applies the transformation to the component.
     */

  }, {
    key: "applyTransformation",
    value: function applyTransformation() {
      if (this._ref) {
        this._ref.style.width = "".concat(this.props.diameter, "px");
        this._ref.style.height = "".concat(this.props.diameter, "px");
        this._ref.style.opacity = 1;
      }
    }
    /**
     * The render function.
     */

  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          animationDuration = _this$props.animationDuration,
          className = _this$props.className,
          diameter = _this$props.diameter,
          children = _this$props.children,
          position = _this$props.position,
          segmentAngles = _this$props.segmentAngles,
          style = _this$props.style,
          passThroughProps = _objectWithoutProperties(_this$props, ["animationDuration", "className", "diameter", "children", "position", "segmentAngles", "style"]);

      var finalClassName = className ? "".concat(className, " ").concat(this.className) : this.className;
      return _react["default"].createElement("div", _extends({
        ref: function ref(i) {
          return _this2._ref = i;
        },
        className: finalClassName,
        style: _objectSpread({
          transition: "all ".concat(animationDuration, "ms"),
          left: position[0] - diameter / 2,
          top: position[1] - diameter / 2
        }, style)
      }, passThroughProps), children.map(function (child, idx, children) {
        var start = segmentAngles[0];
        var end = segmentAngles[1];
        var range = end - start;
        var amount = range > 270 ? children.length : children.length - 1;
        var rotationAngle = start + range / amount * idx;
        return _react["default"].createElement(_CircleMenuItem["default"], {
          radius: diameter / 2,
          rotationAngle: rotationAngle,
          idx: idx,
          animationDuration: _this2.props.animationDuration,
          key: idx
        }, child);
      }));
    }
  }]);

  return CircleMenu;
}(_react["default"].Component);

exports.CircleMenu = CircleMenu;

_defineProperty(CircleMenu, "propTypes", {
  className: _propTypes["default"].string,
  style: _propTypes["default"].object,

  /**
   * The duration of the animation in milliseconds. Pass 0 to avoid animation.
   * Default is 300.
   *
   * @type {Number}
   */
  animationDuration: _propTypes["default"].number,

  /**
   * The diameter of the CircleMenu in pixels. Default is 100.
   *
   * @type {Number}
   */
  diameter: _propTypes["default"].number,

  /**
   * The children of the CircleMenu. Most common are buttons.
   *
   * @type {Node}
   */
  children: _propTypes["default"].node.isRequired,

  /**
   * An array containing the x and y coordinates of the CircleMenus Center.
   * @type {Number[]}
   */
  position: _propTypes["default"].arrayOf(_propTypes["default"].number).isRequired,

  /**
   * Optional Segement of angles where to show the children. Default is [0, 360].
   */
  segmentAngles: _propTypes["default"].arrayOf(_propTypes["default"].number)
});

_defineProperty(CircleMenu, "defaultProps", {
  animationDuration: 300,
  diameter: 100,
  segmentAngles: [0, 360]
});

var _default = CircleMenu;
exports["default"] = _default;
