// @terrestris/react-geo/dist/Map/FloatingMapLogo/FloatingMapLogo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

require("./FloatingMapLogo.less");

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Class representing a floating map logo
 *
 * @class The FloatingMapLogo
 * @extends React.Component
 */
var FloatingMapLogo =
/*#__PURE__*/
function (_React$Component) {
  _inherits(FloatingMapLogo, _React$Component);

  function FloatingMapLogo() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, FloatingMapLogo);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(FloatingMapLogo)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "floatingmaplogo"));

    return _this;
  }

  _createClass(FloatingMapLogo, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          imageSrc = _this$props.imageSrc,
          imageHeight = _this$props.imageHeight,
          absolutelyPositioned = _this$props.absolutelyPositioned,
          className = _this$props.className,
          style = _this$props.style;
      var finalClassName = className ? "".concat(className, " ").concat(this.className) : this.className;

      if (absolutelyPositioned) {
        Object.assign(style, {
          'position': 'absolute'
        });
      }

      return _react["default"].createElement("img", {
        className: finalClassName,
        src: imageSrc,
        height: imageHeight,
        style: style
      });
    }
  }]);

  return FloatingMapLogo;
}(_react["default"].Component);

_defineProperty(FloatingMapLogo, "propTypes", {
  /**
   * An optional CSS class which should be added.
   * @type {String}
   */
  className: _propTypes["default"].string,

  /**
   * The imageSrc (required property).
   * @type {String}
   */
  imageSrc: _propTypes["default"].string.isRequired,

  /**
   * The image height
   * @type {String}
   */
  imageHeight: _propTypes["default"].string,

  /**
   * Whether the map-logo is absolutely postioned or not
   * @type {boolean}
   */
  absolutelyPositioned: _propTypes["default"].bool,

  /**
   * The style object
   * @type {Object}
   */
  style: _propTypes["default"].object
  /**
   * The default properties.
   * @type {Object}
   */

});

_defineProperty(FloatingMapLogo, "defaultProps", {
  imageSrc: 'logo.png',
  absolutelyPositioned: false
  /**
   * The render function.
   */

});

var _default = FloatingMapLogo;
exports["default"] = _default;
