// @terrestris/react-geo/dist/HigherOrderComponent/VisibleComponent/VisibleComponent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isVisibleComponent = isVisibleComponent;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Logger = _interopRequireDefault(require("@terrestris/base-util/dist/Logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The HOC factory function.
 *
 * Wrapped components will be checked against the activeModules array of
 * the state: If the wrapped component (identified by it's name) is included
 * in the state, it will be rendered, if not, it wont.
 *
 * @param {Component} WrappedComponent The component to wrap and enhance.
 * @param {Object} options The options to apply.
 * @return {Component} The wrapped component.
 */
function isVisibleComponent(WrappedComponent) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$withRef = _ref.withRef,
      withRef = _ref$withRef === void 0 ? false : _ref$withRef;

  /**
   * The wrapper class for the given component.
   *
   * @class The VisibleComponent
   * @extends React.Component
   */
  var VisibleComponent =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(VisibleComponent, _React$Component);

    /**
     * The props.
     * @type {Object}
     */
    function VisibleComponent(props) {
      var _this;

      _classCallCheck(this, VisibleComponent);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(VisibleComponent).call(this, props));
      /**
       * The wrapped instance.
       * @type {Element}
       */

      _defineProperty(_assertThisInitialized(_this), "getWrappedInstance", function () {
        if (withRef) {
          return _this.wrappedInstance;
        } else {
          _Logger["default"].debug('No wrapped instance referenced, please call the ' + 'isVisibleComponent with option withRef = true.');
        }
      });

      _defineProperty(_assertThisInitialized(_this), "setWrappedInstance", function (instance) {
        if (withRef) {
          _this.wrappedInstance = instance;
        }
      });

      _defineProperty(_assertThisInitialized(_this), "isVisibleComponent", function (componentName) {
        var activeModules = _this.props.activeModules || [];
        return activeModules.some(function (activeModule) {
          if (!activeModule.name) {
            return false;
          } else {
            return activeModule.name === componentName;
          }
        });
      });

      _this.wrappedInstance = null;
      return _this;
    }
    /**
     * Returns the wrapped instance. Only applicable if withRef is set to true.
     *
     * @return {Element} The wrappend instance.
     */


    _createClass(VisibleComponent, [{
      key: "render",

      /**
       * The render function.
       */
      value: function render() {
        // Filter out extra props that are specific to this HOC and shouldn't be
        // passed through.
        var _this$props = this.props,
            activeModules = _this$props.activeModules,
            passThroughProps = _objectWithoutProperties(_this$props, ["activeModules"]); // Check if the current component should be visible or not.


        var isVisibleComponent = this.isVisibleComponent(passThroughProps.name); // Inject props into the wrapped component. These are usually state
        // values or instance methods.

        return isVisibleComponent ? _react["default"].createElement(WrappedComponent, _extends({
          ref: this.setWrappedInstance
        }, passThroughProps)) : null;
      }
    }]);

    return VisibleComponent;
  }(_react["default"].Component);

  _defineProperty(VisibleComponent, "propTypes", {
    activeModules: _propTypes["default"].arrayOf(_propTypes["default"].object)
    /**
     * Create the VisibleComponent.
     *
     * @constructs VisibleComponent
     */

  });

  return VisibleComponent;
}
